### ISSUE TITLE
[Windows] support pubspec.yaml version/build number in windows projects

### ISSUE THREAD

**Issue by narumi147** [flutter/flutter#73652]
## Steps to Reproduce

1. `flutter config --enable-windows-desktop`
2. create a new project and run `flutter build windows`, the built exe version should be 1.0.0.0
3. change version in `pubspec.yaml` from `1.0.0+1` to any other version like `1.1.1+2`
4.  `pub get` and build again, the version of exe is still `1.0.0.0` 
5. `flutter clean` doesn't help

I have to change the default value of `VERSION_AS_NUMBER` and `VERSION_AS_STRING` in `windows/runner/Runner.rc`. However, these two value **SHOULD** be used only when `FLUTTER_BUILD_NUMBER` and `FLUTTER_BUILD_NAME` not defined, and they should be defined from version in `pubspec.yaml`  when run pub get ?
```
#ifdef FLUTTER_BUILD_NUMBER
#define VERSION_AS_NUMBER FLUTTER_BUILD_NUMBER
#else
#define VERSION_AS_NUMBER 1,1,1,2
#endif

#ifdef FLUTTER_BUILD_NAME
#define VERSION_AS_STRING #FLUTTER_BUILD_NAME
#else
#define VERSION_AS_STRING "1.1.1.2"
#endif
```
<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.26.0-2.0.pre.167, on Microsoft Windows [Version 10.0.19042.685], locale zh-CN)
    • Flutter version 1.26.0-2.0.pre.167 at D:\Projects\sdk\flutter
    • Framework revision d3a2259541 (9 days ago), 2020-12-31 15:14:03 -0500
    • Engine revision 82b4ae86d6
    • Dart version 2.12.0 (build 2.12.0-179.0.dev)
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at D:\Projects\sdk\Android\Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = D:\Projects\sdk\Android\Sdk
    • Java binary at: C:\Users\narumi\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-1\201.6953283\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at .\Google\Chrome\Application\chrome.exe)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Visual Studio - develop for Windows (Visual Studio Enterprise 2019 16.8.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise
    • Visual Studio Enterprise 2019 version 16.8.30804.86
    • Windows 10 SDK version 10.0.18362.0

[✓] Android Studio (version 4.1.0)
    • Android Studio at C:\Users\narumi\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-1\201.6953283
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] VS Code (version 1.52.1)
    • VS Code at C:\Users\narumi\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.19042.685]
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 87.0.664.75

! Doctor found issues in 1 category.
```
</details>

**Comment by stuartmorgan** [comment]
> However, these two value **SHOULD** be used only when `FLUTTER_BUILD_NUMBER` and `FLUTTER_BUILD_NAME` not defined, and they should be defined from version in `pubspec.yaml` when run pub get ?

Setting those defines from pubspec.yaml has never actually been implemented; that code is there to future-proof projects so that it can be added later and existing projects will benefit automatically.

I thought I filled an issue to track implementing that piece when landing the template, but I don't see it, so we can use this one.

**Comment by fzyzcjy** [comment]
A year has passed... Does this mean that, we should *duplicate* writing versions?

**Comment by lewcianci** [comment]
It's a little hard to believe that Flutter for Windows went stable without this critical feature...

**Comment by Sunbreak** [comment]
Try flutter_distributor according to https://github.com/leanflutter/flutter_distributor/issues/65

**Comment by stuartmorgan** [comment]
> It's a little hard to believe that Flutter for Windows went stable without this critical feature...

It is a one-line change in `Runner.rc` to update the build number. Needing to change the version number in two places instead of one certainly isn't ideal, but "critical feature" substantially overstates the severity here.

**Comment by jmatth** [comment]
Here are the Github Actions steps I'm using to replace the build number as part of the Windows build process:

```yaml
      - name: Extract version from pubspec
        shell: bash
        run: |
          version="$(grep '^version:\s\+[[:digit:]]\+\.[[:digit:]]\+\.[[:digit:]]\+$' pubspec.yaml | head -n1 | awk '{print $2}')"
          if test -z "$version"; then
            echo 'ERROR: failed to get version from pubspec.yaml'
            exit 1
          fi
          echo "RELEASE_VERSION=$version" >> $GITHUB_ENV
          echo "RELEASE_VERSION_NUM=${version//./,}" >> $GITHUB_ENV

      # Several unrelated steps omitted

      - name: Fix app version (Windows)
        if: runner.os == 'windows'
        run: |
          (Get-Content .\windows\runner\Runner.rc) `
          -replace 'VERSION_AS_NUMBER 1,0,0', 'VERSION_AS_NUMBER ${{ env.RELEASE_VERSION_NUM }}' `
          -replace 'VERSION_AS_STRING "1.0.0"', 'VERSION_AS_STRING "${{ env.RELEASE_VERSION }}"' `
          | Out-File -Encoding UTF8 .\windows\runner\Runner.rc
```

Note that it assumes the version number extracted from `pubspec.yaml` is in the form `x.y.z`. You'll probably need more logic to sanitize the value for `VERSION_AS_NUMBER` if you want to use values like `x.y.z+a`.

**Comment by lewcianci** [comment]
> > It's a little hard to believe that Flutter for Windows went stable without this critical feature...
> 
> It is a one-line change in `Runner.rc` to update the build number. Needing to change the version number in two places instead of one certainly isn't ideal, but "critical feature" substantially overstates the severity here.

I think it's really important. To that end, would you accept a PR to resolve the issue?

**Comment by stuartmorgan** [comment]
> To that end, would you accept a PR to resolve the issue?

We're always happy to accept high-quality PRs to resolve issues.

> And, is there somewhere in the Flutter repo you could direct me to, to where this verison is generated?

I'm not sure what you mean by "generated", since the desired behavior is to use the version that's already in `pubspec.yaml`. If you mean where build defines should be set, https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/windows/build_windows.dart is the primary driver of the Windows build.

> so I'm not sure why it's aceptable for the Windows platform

I'm not sure what "acceptable" means here. It's a known issue, tracked in the issue tracker.

**Comment by lewcianci** [comment]
> > so I'm not sure why it's aceptable for the Windows platform
> 
> I'm not sure what "acceptable" means here. It's a known issue, tracked in the issue tracker.

I mean that it wouldn't have been acceptable for Flutter to hit stable on Android or iOS if it just versioned APK's and IPA's at 1.0.0, and you had to manually bump the versions yourself. The first part of the quoted sentence covered that. Versioning assemblies is a huge deal - how can Flutter for desktop be ready for production use if every Windows app you produce is going to be 1.0.0 unless you manually fudge the build files? How can you hope to ever update a published app if your published app's assembly is always at 1.0.0?

It's this disparity in functionality between certain stable platforms, like Android/iOS, and other stable platforms, like Windows, that's preventing Windows from feeling like a first-class development experience on the platform. Windows support still feels like beta, despite being announced as stable, due to weird quirks like this.

Anyway, me musing about this isn't getting it fixed or actually progressing the issue here, so I'll stop complaining and instead try to get a PR together. I'm not trying to riff on the Flutter team's efforts, nor you personally @stuartmorgan, I just feel like the ability to version an assembly is important in understanding crash statistics and also handling application updates. I'm still thoroughly enthralled with Flutter and intend to use it for years to come.

**Comment by stuartmorgan** [comment]
> unless you manually fudge the build files?

`.rc` files are source files, not build files. There is no "fudging of build files" happening here, any more than changing `pubspec.yaml` is.

> How can you hope to ever update a published app if your published app's assembly is always at 1.0.0?

Again, the issue here is not "is impossible to change the Windows application's version" it is "when changing your application's version, you need to do it in one extra place".

There are a number of cases in Flutter development where is necessary to work directly with the platform project to change metadata (display name, icon, permissions, etc.). Doing so is neither impossible, nor unprecedented, for Flutter.

On Windows, that currently includes version. It would certainly be better if that were not the case, but framing the issue as it being impossible to release updates is not constructive.

> It's this disparity in functionality between certain stable platforms, like Android/iOS, and other stable platforms, like Windows, that's preventing Windows from feeling like a first-class development experience on the platform.

Rewatching stable is a milestone, not an end state; Windows support will continue to improve. Android and iOS have been in stable for several years, Windows only reached stable a few months ago; some disparity is inevitable. I suspect that if you tried switching to Flutter 1.0 you would have a lot of complaints and Android and iOS support relative to its current state.

**Comment by lewcianci** [comment]
You know what, you're right, I'm wrong. I didn't understand the role that `.rc` files played, being comparatively new to C++ and CMAKE projects (only really getting into them for Flutter on Windows development). You're especially right about the Flutter Android/iOS 1.0 statement - I would probably not be happy with aspects of those either. I am blowing this out of proportion.

At any rate, thanks for taking the time to debate (and reason with) some random developer from the other side of the world about this. One of the things that I think is the best about Flutter is the community, and I think that in quite a few other repo's I would just be told to pull my head in. But the fact you, someone with intimate knowledge of Flutter and probably very educated, took the time to explain this to me and also reason with me helps me reframe how I see it as well. You're not here to necessarily convince anyone of anything, but the fact that you took the time to do so is something I appreciate immensely.

Thanks for being so awesome @stuartmorgan.

**Comment by Merrit** [comment]
Excellent! :tada:

Thank you for all your hard work :bowing_woman:

**Comment by ollydixon** [comment]
@loic-sharma thanks for this, but what if we have the message ```Invalid project name: 'flutter' - this will conflict with Flutter package dependencies.```

**Comment by ollydixon** [comment]
@loic-sharma it seems to delete files like pubspec.yaml and main.dart in the existing project.

**Comment by loic-sharma** [comment]
@ollydixon I wouldn't update your project yet as this feature isn't released. That said, what deleted your pubspec.yaml and main.dart files? Running `flutter create` should not delete or modify files that already exist.

**Comment by loic-sharma** [comment]
Added in https://github.com/flutter/flutter/pull/106145 and will be available in the next stable release of Flutter.

**Comment by stuartmorgan** [comment]
> Running flutter create even with the target resets the core files

@ollydixon If you have a reproducible case where `flutter create` deletes or modifies existing files, please file an issue with details.

**Comment by loic-sharma** [comment]
Hello everyone, this feature was released today with Flutter 3.3: https://docs.flutter.dev/deployment/windows#updating-the-apps-version-number

Windows projects that were created before Flutter 3.3 will need to be migrated: https://docs.flutter.dev/development/platform-integration/windows/version-migration

If you run into any problems, feel free to open a new issue!
